src/pages/docs/ai-transport/features/token-streaming/message-per-token.mdx

```diff
@@ -0,0 +1,429 @@
+
+<Code>
+```javascript
+const channel = realtime.channels.get('{{RANDOM_CHANNEL_NAME}}');
+
+// Track responses by ID
+const responses = new Map();
+
+await channel.subscribe('token', (message) => {
+  const token = message.data;
+  const responseId = message.extras?.headers?.responseId;
+
+  if (!responseId) {
+    console.warn('Token missing responseId');
+    return;
+  }
+
+  // Create an empty response
+  if (!responses.has(responseId)) {
+    responses.set(responseId, '');
+  }
+
+  // Append token to response
+  responses.set(responseId, responses.get(responseId) + token);
+});
+```
+</Code>
+
+### Token stream with explicit start/stop events <a id="explicit-events"/>
+
+In some cases, your AI model response stream may include explicit events to mark response boundaries. You can indicate the event type, such as a response start/stop event, using the Ably message name.
+
+#### Publish tokens
+
+<Code>
+```javascript
+const channel = realtime.channels.get('{{RANDOM_CHANNEL_NAME}}');
+
+// Example: stream returns events like:
+// { type: 'message_start', responseId: 'resp_abc123' }
+// { type: 'message_delta', responseId: 'resp_abc123', text: 'Hello' }
+// { type: 'message_stop', responseId: 'resp_abc123' }
+
+for await (const event of stream) {
+  if (event.type === 'message_start') {
+    // Publish response start
+    channel.publish({
+      name: 'start',
+      extras: {
+        headers: {
+          responseId: event.responseId
+        }
+      }
+    });
+  } else if (event.type === 'message_delta') {
+    // Publish tokens
+    channel.publish({
+      name: 'token',
+      data: event.text,
+      extras: {
+        headers: {
+          responseId: event.responseId
+        }
+      }
+    });
+  } else if (event.type === 'message_stop') {
+    // Publish response stop
+    channel.publish({
+      name: 'stop',
+      extras: {
+        headers: {
+          responseId: event.responseId
+        }
+      }
+    });
+  }
+}
+```
+</Code>
+
+#### Subscribe to tokens
+
+Handle each event type to manage response lifecycle:
+
+<Code>
+```javascript
+const channel = realtime.channels.get('{{RANDOM_CHANNEL_NAME}}');
+
+const responses = new Map();
+
+// Handle response start
+await channel.subscribe('start', (message) => {
+  const responseId = message.extras?.headers?.responseId;
+  responses.set(responseId, '');
+});
+
+// Handle tokens
+await channel.subscribe('token', (message) => {
+  const responseId = message.extras?.headers?.responseId;
+  const token = message.data;
+
+  const currentText = responses.get(responseId) || '';
+  responses.set(responseId, currentText + token);
+});
+
+// Handle response stop
+await channel.subscribe('stop', (message) => {
+  const responseId = message.extras?.headers?.responseId;
+  const finalText = responses.get(responseId);
```

Comment:
Perhaps (assuming that the idea of the `responses` map is just to accumulate response content during generation) remove from `responses`?